File: day25/mergeIntervals.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstdlib>
#include <ctime>

using Interval = std::pair<int, int>;

std::vector<Interval> mergeIntervals(std::vector<Interval>& intervals) {
    if (intervals.empty()) return {};

    std::sort(intervals.begin(), intervals.end());

    std::vector<Interval> merged;
    merged.push_back(intervals[0]);

    for (size_t i = 1; i < intervals.size(); ++i) {
        Interval& last = merged.back();
        if (intervals[i].first <= last.second) {
            last.second = std::max(last.second, intervals[i].second);
        } else {
            merged.push_back(intervals[i]);
        }
    }

    return merged;
}

std::vector<Interval> generateTestCase() {
    std::srand(static_cast<unsigned>(std::time(nullptr)));
    int n = std::rand() % 6 + 5; // 5–10 intervals
    std::vector<Interval> intervals;
    for (int i = 0; i < n; ++i) {
        int start = std::rand() % 21;
        int end = start + std::rand() % 10 + 1;
        intervals.emplace_back(start, end);
    }
    std::random_shuffle(intervals.begin(), intervals.end());
    return intervals;
}

void printIntervals(const std::vector<Interval>& intervals) {
    for (const auto& [start, end] : intervals) {
        std::cout << "[" << start << "," << end << "] ";
    }
    std::cout << "\n";
}

int main() {
    std::vector<Interval> intervals = generateTestCase();

    std::cout << "Input: ";
    printIntervals(intervals);

    std::vector<Interval> merged = mergeIntervals(intervals);

    std::cout << "Merged: ";
    printIntervals(merged);

    return 0;
}

File: day25/merge_intervals.rs
use rand::{random_range, rng};
use rand::seq::SliceRandom;

fn merge_intervals(mut intervals: Vec<(i32, i32)>) -> Vec<(i32, i32)> {
    if intervals.is_empty() {
        return vec![];
    }

    intervals.sort_by_key(|x| x.0);
    let mut merged = vec![intervals[0]];

    for current in intervals.into_iter().skip(1) {
        let last = merged.last_mut().unwrap();
        if current.0 <= last.1 {
            last.1 = last.1.max(current.1);
        } else {
            merged.push(current);
        }
    }

    merged
}

fn generate_test_case() -> Vec<(i32, i32)> {
    let len = random_range(5..=10);
    let mut intervals = Vec::with_capacity(len);
    for _ in 0..len {
        let start = random_range(0..=20);
        let end = start + random_range(1..=10);
        intervals.push((start, end));
    }
    intervals.shuffle(&mut rng());
    intervals
}

fn main() {
    let intervals = generate_test_case();
    println!("Input: {:?}", intervals);
    let merged = merge_intervals(intervals);
    println!("Merged: {:?}", merged);
}
